### PR TITLE
Refactor of rz-test to use thread safe functions.

### PR DESCRIPTION
**Pull request by wargio**
# do not squash
 

**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

This uses the thread-safe functions and cleans up a bit the rz-test code.

Now we automatically detect the core number and use them all unless `-j <n>` is specified.

### CODE REVIEW

File: binrz/rz-test/rz-test.c [hunk 1]
```diff
@@ -26,36 +25,39 @@
...
-	ut64 xx;
-	ut64 br;
-	ut64 fx;
-} RzTestFileCounts;
+typedef struct test_counter_t {
+	size_t n_tests; ///< Total number of tests
+	size_t n_handled; ///< Tests that has been handled
+	size_t ok; ///< Tests that succeeded
+	size_t xx; ///< Tests that failed
+	size_t fx; ///< Tests that succeeded but als marked as broken
```
Comment (Rot127) [inline binrz/rz-test/rz-test.c]:
```suggestion
	size_t fx; ///< Tests that succeeded but also marked as broken
```

File: binrz/rz-test/run.c [hunk 2]
```diff
@@ -558,7 +558,7 @@ RZ_API RzTestResultInfo *rz_test_run_test(RzTestRunConfig *config, RzTest *test)
 	return ret;
 }
 
-RZ_API void rz_test_test_result_info_free(RzTestResultInfo *result) {
+RZ_API void rz_test_result_info_free(RzTestResultInfo *result) {
```
Comment (XVilka) [inline binrz/rz-test/run.c]: Add `RZ_NULLABLE` while you are on it

File: binrz/rz-test/rz_test.h [hunk 3]
```diff
@@ -213,6 +213,6 @@ RZ_API void rz_test_test_free(RzTest *test);
 RZ_API char *rz_test_test_name(RzTest *test);
 RZ_API bool rz_test_test_broken(RzTest *test);
 RZ_API RzTestResultInfo *rz_test_run_test(RzTestRunConfig *config, RzTest *test);
-RZ_API void rz_test_test_result_info_free(RzTestResultInfo *result);
+RZ_API void rz_test_result_info_free(RzTestResultInfo *result);
```
Comment (XVilka) [inline binrz/rz-test/rz_test.h]: Same

File: librz/util/subprocess.c [hunk 4]
```diff
@@ -47,21 +47,25 @@ struct rz_subprocess_t {
...
+	RzThreadLock *subproc_mutex;
+	long refcount;
+	bool has_procthreadattr;
+	volatile long pipe_id;
+	DWORD mode_stdin;
+	DWORD mode_stdout;
+	DWORD mode_stderr;
+} SubprocessWindows;
+
+static SubprocessWindows subwin = { 0 };
```
Comment (XVilka) [inline librz/util/subprocess.c:61]: Add a comment that it is global/static on purpose

File: librz/util/subprocess.c [hunk 5]
```diff
@@ -746,28 +750,32 @@ struct rz_subprocess_t {
...
-static int sigchld_pipe[2];
-static RzThread *sigchld_thread;
+typedef struct subprocess_unix_t {
+	RzPVector /*<RzSubprocess *>*/ subprocs;
+	RzThreadLock *subprocs_mutex;
+	int sigchld_pipe[2];
+	RzThread *sigchld_thread;
+} SubprocessUnix;
+
+static SubprocessUnix subnix = { 0 };
```
Comment (XVilka) [inline librz/util/subprocess.c:762]: Same

**Review by XVilka — CHANGES_REQUESTED** [review]
Nice cleanup/modernization!

`rz-bindgen` returns error:
```
 [1/7] Generating bindgen_outputs with a custom command
FAILED: rizin.i 
/usr/bin/python3 ../src/main.py -o . --clang-path /usr/lib/llvm-14/lib --clang-args ' -resource-dir="/usr/lib/llvm-14/lib/clang/14.0.0"' --rizin-include-path /usr/include/librz --targets SWIG
Traceback (most recent call last):
  File "/home/runner/work/rizin/rizin/rz-bindgen/build/../src/main.py", line 51, in <module>
    bindings.run()
  File "/home/runner/work/rizin/rizin/rz-bindgen/src/bindings.py", line 43, in run
    builders = [HeaderBuilder(name) for name in threaded_headers]
  File "/home/runner/work/rizin/rizin/rz-bindgen/src/bindings.py", line 43, in <listcomp>
    builders = [HeaderBuilder(name) for name in threaded_headers]
  File "/home/runner/work/rizin/rizin/rz-bindgen/src/cparser_header.py", line 101, in __init__
    assert os.path.exists(filename)
AssertionError
[2/7] Generating plugin/swig_runtime with a custom command
```
The macOS unit test error is now clearer, nice:
```
>> MALLOC_PERTURB_=178 /private/var/folders/bc/yvvw2g2s4jlg9r8x7tf61xv800009d/T/woodpecker-local-2757858299/workspace/build/test/unit/test_core_task
――――――――――――――――――――――――――――――――――――― ✀  ―――――――――――――――――――――――――――――――――――――
Hello
this
is
the
main
task!
test_core_task ERR
[XX] Fail at line 44: fcn result: expected 1337, 0
1337, 1
1337, 2
1337, 3
1337, 4
, got (null).

test_core_task_finished_cb ERR
[XX] Fail at line 73: task joined: expected true, got false
```

**Comment by wargio** [comment]
The unit tests fails are unrelated to this PR. check dev

**Comment by XVilka** [comment]
> The unit tests fails are unrelated to this PR. check dev

I know they are unrelated, I thought the error improved after this one, but it apparently happended in dev already. So yes, ignore that part of the comment.

**Comment by XVilka** [comment]
I think having overall count is useful though, could you please also add a column with all tests ran?

Also maybe a final report, how many OK, BR, XX are at the end total?

**Comment by wargio** [comment]
> I think having overall count is useful though, could you please also add a column with all tests ran?

i have removed it by mistake. i changed the code and now it prints it.

**Comment by wargio** [comment]
The issue with Bindgen is unrelated to this PR
```
AssertionError: File /usr/include/librz/sdb/ls.h does not exist
```

**Comment by XVilka** [comment]
> The issue with Bindgen is unrelated to this PR
> 
> ```
> AssertionError: File /usr/include/librz/sdb/ls.h does not exist
> ```

https://github.com/rizinorg/rz-bindgen/pull/64

**Review by XVilka — APPROVED** [review]
There is still room for improvement, e.g. make all XX red in the output if not zero, BR - yellow if not zero, etc. But this is for separate PR.

**Review by kazarmy — COMMENTED** [review]
A running total would be nice -- I generally use it to see how far the build has progressed through the tests. Not sure whether it's strictly necessary, or how to integrate that info in a clean manner.

**Comment by wargio** [comment]
> A running total would be nice -- I generally use it to see how far the build has progressed through the tests. Not sure whether it's strictly necessary, or how to integrate that info in a clean manner.

The running total is printed at the end when `-L` is used.
```
[21179/21179]    20129 OK      1016 BR        0 XX       34 FX
```

**Pull request merged**